### PR TITLE
fix(install): stage query downloads

### PR DESCRIPTION
**Pull request by atusy**
## Summary
- stage query downloads in a temp dir and publish only after the required query file succeeds
- preserve valid legacy query dirs, repair zero-byte/partial installs, and keep failed --force reinstalls from deleting working queries
- recover kakehashi-owned interrupted query backups, keep status/uninstall scans from treating internal dirs as languages, and remove backups under the install lock during uninstall

## Tests
- cargo test install_repairs_partial_query_dir_without_force --lib -- --nocapture
- cargo test install_treats_marker_with_empty_highlights_as_complete --lib -- --nocapture
- cargo test --test test_cli test_language_uninstall_removes_query_backups -- --nocapture
- cargo test --test test_cli test_language_status_zero_byte_unmarked_queries_missing -- --nocapture
- make check

Fixes part of #588.

## Summary by CodeRabbit

* **New Features**
  * Safer query installation via staging, completion markers, interruption-safe recovery, and atomic replacement.
  * Automatic recovery of interrupted query installs by restoring stranded generated backups when appropriate.
  * CLI install/uninstall flow now clears stale uninstall markers before installing queries.
* **Bug Fixes**
  * `language status` ignores dot-prefixed internal query folders and treats zero-byte/incomplete queries as not installed.
  * `language uninstall` now fails when query/parser removal fails and handles hidden internal backups correctly.
* **Tests**
  * Expanded CLI integration coverage for recovery, completeness detection, dot-prefixed directories, and uninstall edge cases.
* **Chores**
  * Added a new direct filesystem helper dependency.

### CODE REVIEW

**Comment by atusy** [comment]
@copilot review

**Comment by atusy** [comment]
@gemini-code-assist review

File: src/install/queries.rs [hunk 1]
```diff
@@ -383,33 +895,61 @@ mod tests {
...
+
         base_url
     }
 
+    fn dead_test_pid() -> u32 {
+        let mut pid = std::process::id().saturating_add(100_000);
+        while process_is_running(&pid.to_string()) {
+            pid = pid.saturating_add(1);
+        }
+        pid
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:1001]: Fixed in fee8334b4 — dead_test_pid and both tests that depend on real process liveness are now #[cfg(unix)]: the non-unix process_is_running conservatively reports every numeric pid live, which would spin the loop forever.

File: src/install/queries.rs [hunk 2]
```diff
@@ -237,9 +266,14 @@ fn install_queries_recursive(
     }
 
     let queries_dir = data_dir.join("queries").join(language);
-
-    // Check if queries already exist
-    if queries_dir.exists() && !force {
+    let queries_parent = data_dir.join("queries");
+    fs::create_dir_all(&queries_parent)?;
+    recover_interrupted_query_install(&queries_parent, language)?;
+
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:282]: Fixed in fee8334b4 — a TempQueryDirGuard (RAII) removes the staging dir on drop, replacing the per-path manual cleanups and covering any future `?` propagation; on success the dir is renamed away first, making the drop a no-op.

File: src/bin/main.rs [hunk 3]
```diff
@@ -683,6 +701,11 @@ fn run_install(language: &str, force: bool, verbose: bool, no_cache: bool) -> Re
     let mut parser_success = true;
     let mut queries_success = true;
 
+    if let Err(e) = queries::clear_uninstall_tombstone_for_install(&data_dir, language) {
+        eprintln!("✗ Failed to prepare query installation: {}", e);
+        queries_success = false;
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs:717]: Refuting: clear_uninstall_tombstone only errs when a tombstone file EXISTS and cannot be removed (NotFound maps to Ok). A surviving tombstone marks the language's queries uninstalled at runtime, so even a subsequently successful download is unusable — the exit code must signal that. Pre-failing is therefore the honest outcome, not an over-eager one; the install itself still proceeds and its own result is reported independently.

File: src/install/queries.rs [hunk 4]
```diff
@@ -352,6 +404,472 @@ fn install_queries_recursive(
...
+        return Ok(());
+    }
+    if queries_dir.exists() {
+        return Ok(());
+    }
+
+    let ownership = backup_ownership_sidecar(&backup_dir);
+    fs::rename(&backup_dir, queries_dir)?;
+    let _ = fs::remove_file(ownership);
+    Ok(())
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:707]: Fixed in a426bc78a — the backup is selected under the replace lock, and a NotFound rename (external cleanup the lock cannot serialize) is treated as nothing-to-restore instead of an error.

File: src/bin/main.rs [hunk 5]
```diff
@@ -547,6 +558,7 @@ fn run_language_uninstall(
 
     // Uninstall each language
     let mut any_removed = false;
+    let mut any_failed = false;
     for lang in &languages_to_uninstall {
         let mut removed_something = false;
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/bin/main.rs:573]: Fixed in 07ec14392 — the uninstall loop now rejects names failing `is_safe_language_name` before any path is built (parser removal included), printing the name debug-escaped and failing the run.

File: src/install/queries.rs [hunk 6]
```diff
@@ -352,6 +404,481 @@ fn install_queries_recursive(
...
+        }
+        if let Some(language) = backup_language_name(&path) {
+            recover_interrupted_query_install(queries_parent, &language)?;
+        } else if let Some((language, _)) = temp_language_name_and_pid(&path) {
+            remove_interrupted_temp_query_install(queries_parent, &language, &path)?;
+        }
+    }
+
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:505]: Fixed in 07ec14392 — `recover_interrupted_query_installs` now dedupes by language via a HashSet, so recovery (and its rescan + lock acquisition) runs at most once per language per scan.

File: src/install/queries.rs [hunk 7]
```diff
@@ -352,6 +404,481 @@ fn install_queries_recursive(
...
+        fs::rename(tmp_queries_dir, queries_dir)?;
+        return Ok(ReplaceQueryDirResult::Replaced);
+    }
+
+    let backup_dir = unique_backup_query_dir(queries_dir, language);
+    write_backup_ownership_marker(&backup_dir)?;
+    if let Err(e) = fs::rename(queries_dir, &backup_dir) {
+        let _ = fs::remove_file(backup_ownership_sidecar(&backup_dir));
+        return Err(QueryInstallError::IoError(e));
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:874]: Fixed in 4fcf11008 — a NotFound from the backup rename now publishes the staged dir (rename tmp -> canonical) and returns Replaced instead of aborting; the pre-created sidecar is removed on that path as before.

File: src/install/queries.rs [hunk 8]
```diff
@@ -352,6 +404,481 @@ fn install_queries_recursive(
...
+    };
+
+    let ownership = backup_ownership_sidecar(&backup_dir);
+    match fs::rename(&backup_dir, queries_dir) {
+        Ok(()) => {}
+        // The backup vanished after selection (external cleanup — the lock
+        // only serializes kakehashi's own installers): nothing to restore.
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Ok(()),
+        Err(e) => return Err(QueryInstallError::IoError(e)),
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:705]: Fixed in 4fcf11008 — the NotFound branch now removes the orphaned ownership sidecar (best-effort, idempotent) before returning Ok.

File: src/install/queries.rs [hunk 9]
```diff
@@ -383,33 +917,62 @@ mod tests {
...
-            let mut reader = BufReader::new(&mut stream);
-            let mut line = String::new();
-            let _ = reader.read_line(&mut line);
-            loop {
-                line.clear();
-                match reader.read_line(&mut line) {
-                    Ok(0) | Err(_) => break,
-                    Ok(_) if line == "\r\n" || line == "\n" => break,
-                    Ok(_) => {}
+            for stream in listener.incoming() {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs]: Fixed in 09101b2e5 — the accept loop is bounded (incoming().take(64)), so the helper thread and socket terminate instead of outliving the test.

File: src/install/queries.rs [hunk 10]
```diff
@@ -352,6 +404,501 @@ fn install_queries_recursive(
...
+        fs::remove_dir_all(&queries_dir)?;
+        removal.removed_queries = true;
+    }
+
+    let entries = fs::read_dir(queries_parent)?;
+    for entry in entries.flatten() {
+        let path = entry.path();
+        let Some(name) = path.file_name().and_then(|name| name.to_str()) else {
+            continue;
+        };
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs]: Fixed in a8b840e19 — the loop now propagates per-entry read_dir errors (entry?) instead of flatten(), so uninstall fails loudly when it cannot enumerate backups.

File: src/install/queries.rs [hunk 11]
```diff
@@ -87,13 +95,23 @@ pub struct QueryInstallResult {
...
 
+fn validate_safe_language_name(language: &str) -> Result<(), QueryInstallError> {
+    if is_safe_language_name(language) {
+        Ok(())
+    } else {
+        Err(QueryInstallError::LanguageNotSupported(
+            language.escape_default().to_string(),
+        ))
+    }
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:123]: Fixed in a8b840e19 — added QueryInstallError::InvalidLanguageName with its own message; validate_safe_language_name now returns it, and the unsafe-name tests assert the new variant.

File: src/install/queries.rs [hunk 12]
```diff
@@ -352,6 +404,501 @@ fn install_queries_recursive(
...
+        if path.is_dir()
+            && generated_backup_matches_language(name, language)
+            && backup_is_owned(&path)
+        {
+            let ownership = backup_ownership_sidecar(&path);
+            fs::remove_dir_all(path)?;
+            let _ = fs::remove_file(ownership);
+            removal.removed_backups = true;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/install/queries.rs:554]: This landed after the merge — fixed as follow-up PR #678 (NotFound from remove_dir_all treated as already-removed for the canonical dir and owned backups).
